Additional texture modifiers

Treer opened the pull request:
Adds new texture modifiers to perform the following:
* Adjust hue, saturation, and lightness
* Colorize using hue, saturation, and lightness
* Adjust contrast & brightness
* Hard light
* Overlay
* Screen
* Create texture of a given size and color (added in extra commit - can be removed)

More power in the texture modifiers is better, but a use-case example: the ability to adjust and repurpose existing textures instead of a mod providing its own textures allows mods to be compatible with whichever texture pack the player/game is using, and if source textures are obtained from the registered nodes or aliases the mod can even become independent-ish of the game and other default mods - while still providing items and nodes in the same art style and resolution.

The ability to adjust or repurpose textures is extremely limited at the moment. There's no way to change colours or perform these operations, the existing `[colorize` fades a texture out to a colour, `[multiply` can only use colour components to the degree that they already existed in the texture, etc.

These new blend modes and colour tools should be both powerful and comfortable for people familiar with graphics programs like GIMP or Photoshop or mix-blend-mode in CSS. For anyone not already familiar, a short explanation for each is provided in the lua_api.txt part of this commit.

* Resolves #9055
* Some of these modifiers take many parameters simply because it's practical / ease of use, but that may also help with an issue of long modifier pipelines leaving textures from intermediate steps in the cache, by virtue of being able to accomplish multiple tasks in a single texture modifier. (I don't think that issue has a ticket, I just saw it mentioned a few times while looking for related PRs & tickets, though it's possible I misunderstand it)

## To do

This PR is Ready for Review.
<!-- ^ delete one -->

- [x] A [demonstration mod](https://github.com/Treer/texturemodifiertest) to aid testing
- [x] The new Screen modifier is a complement to the existing Multiply modifier, so it only takes a colour as an argument and can't screen blend two textures. I'll try to confirm that Screen and Multiply can now be performed between textures by `A.png^[contrast:0:-64^[overlay:B.png` and `A.png^[contrast:0:64^[overlay:B.png`, and if testing shows that to work flawlessly then I'll add the technique to the Screen and Multiply entries in lua_api.txt

## How to test

There is a [demo mod](https://github.com/Treer/texturemodifiertest) which creates 291 nodes covering edge cases, extremes and normal uses of the hsl, colorizehsl, contrast, overlay, hardlight, and screen texture modifiers.

With each node being a uniquely generated 736x736 texture, it's something of a speed/memory test too.

An example texture modifier for a mod with a `flowers` dependency 

    minetest.register_node("texturemodtest:fairyshroom", {
    	description = "Fairyshroom",
    	tiles = {"flowers_mushroom_red.png^[hsl:-123"},
    	inventory_image = "flowers_mushroom_red.png^[hsl:-123",
    	wield_image = "flowers_mushroom_red.png^[hsl:-123",
    	drawtype = "plantlike",
    	paramtype = "light",
    	sunlight_propagates = true,
    	walkable = false,
    	buildable_to = true,
    	groups = {mushroom = 1, food_mushroom = 1, snappy = 3, attached_node = 1, flammable = 1},
    })

creates a fairyshroom that works across existing art styles and resolutions:
![fairyshroom](https://user-images.githubusercontent.com/6390507/85763968-84b9de00-b758-11ea-90d3-55f1a4cf1233.png)![fairyshroom3](https://user-images.githubusercontent.com/6390507/85766506-72d93a80-b75a-11ea-95c9-10444ec010ca.png)

Treer committed:
Additional texture modifiers

Add texture modifiers to perform the following:
Adjust hue-saturation-lightness.
Colorize using hue-saturation-lightness.
Adjust Contrast & Brightness
Hard light
Overlay
Screen.

Extex101 commented:
We need this! :+1: 
P.S. congrats on 10100th issue/PR :P

Treer commented on src/client/tile.cpp:
```diff
@@ -1978,6 +2117,218 @@ static void apply_multiplication(video::IImage *dst, v2u32 dst_pos, v2u32 size,
+	double slope = 1 - std::fabs(norm_b);
+	
+	// Apply the user's contrast adjustment to the calculated slope, such that 
+	// -255 will make it near-vertical and +255 will make it horizontal
```
this comment should read
// -127 will make it near-vertical and +127 will make it horizontal
It will be updated

NathanSalapat commented:
Hue adjustment would be great. I have a few mods that could really use that.

HybridDog commented:
Are the textures changed in a gamma-correct way or are the sRGB values changed directly without gamma correction?

hecktest commented:
Very nice.

Treer commented:
@HybridDog , short answer no, long answer....

The math is linear on the "perceptually uniform RGB" values (which are not linear). I don't know the common term for the other way - I call it "linear light" - but I read "gamma-correct way" as you referring to converting the colour to linear light values before adding or subtracting components.

My initial reaction to the question was "meh, it doesn't matter for these operations", as I have linear light pegged as being important when you're quantizing or summing colour components - most prominently scaling and antialiasing.

My second thought was "hmm, I'm not experienced enough to know whether it matters for these operations". While googling for what the hell the common terminology for this is, I found [this article](https://discuss.pixls.us/t/users-guide-to-high-bit-depth-gimp-2-9-2-part-2/534)

> GIMP 2.9.2 editing operations that automatically linearize the RGB channel values include scaling the image, Gaussian blur, UnSharp Mask, Channel Mixer, Auto Stretch Contrast, decomposing to LAB and LCH, all of the LCH blend modes, and quite a few other editing operations.
>
> GIMP 2.9.2 editing operations that ought to, but don’t yet, linearize the RGB channels include the all-important Curves and Levels operations. 

Which teaches me that my version of GIMP is a fossil and 2.9.x offers important improvements, but also isn't indicating that the texture modifier operations in this PR need to be performed in linear light.

Contrast/Brightness results would change, but that doesn't adjust colour so I don't think either way would be wrong for Contrast/Brightness in the way that antialiasing without linear light is wrong.

So my intuition is it doesn't matter for this purpose, but I'm not qualified to make that claim.

Treer commented:
follow up...

I applied Multiply, screen, brightness, contrast, and hue to a test image in GIMP 2.8, then upgraded to GIMP 2.10 - which is linear light aware - and performed the same operations again. I judged whether they have changed by eye, flicking back and fourth between new and old images.

You can view the results here: [testimage_and_mod.zip](https://github.com/minetest/minetest/files/4840074/testimage_and_mod.zip) (10 MB zip of .xcf and multiply/screen demo mod)


Surprisingly, brightness and contrast did not change. GIMP's Brightness-contrast tool is really just simpler knobs for operating its Curves tool, and the excerpt in my previous comment said GIMP 2.9.2 was not using linear light for Curves, even though it should. So perhaps GIMP 2.10.20 is still not using linear light for Brightness-Contrast and that's why it hasn't changed.

Multiply is different in the darker areas, but I could not see a change in Screen. This is odd because multiply and screen are basically opposites (screen is invert -> multiply -> invert), but perhaps it is due to how linear light makes its noticeable differences in the darks and screen is a blend that makes things lighter?

There was a slight change in the hue operation, none at 180° but a slight change at -135°. I can't tell if that's just differences in precision/rounding or whether they now have a linear light version of HSL.

I haven't discovered anything that makes me concerned about the gamma.

---

In other news I confirmed that the trick mentioned in the "To do" section works, and so you can do a multiply blend between two images with `A.png^[contrast:0:-64^[overlay:B.png` and a screen blend with `A.png^[contrast:0:64^[overlay:B.png`

![screenshot](https://user-images.githubusercontent.com/6390507/85916731-d9637300-b896-11ea-995d-4b3da9945377.jpg)

So I'll add that technique the lua_api.txt part of this PR.

I think I'll also change the Brightness/Contrast modifier in this PR so that the brighness value of -127 to +127 gets scaled to -127.5 to +127.5, because the calculation assumes -127.5 to +127.5, which leads to full brightness being 254 instead of 255 etc

Treer committed:
increase brightness output range

Treer committed:
document a trick for doing multiply/screen blends between textures

HybridDog commented:
> Contrast/Brightness results would change, and this might be worth exploring, but it doesn't change colour so I don't think either way would be bad for Contrast/Brightness in the sense it's bad to antialias without linear light.

Colour is unaffected, but changing the brightness in linear or sRGB colourspace can make a big difference. For example, if the brightness of white is halved, the linear brightness value becomes `(1*0.5)^2.4 = 0.189` instead of `0.5`, which is a significant difference.
When downscaling images correctly (pictures of text or function plots are exceptions), the average linear brightness should stay constant, see for example #6867.
On the other hand, sRGB brightness is perceived linearly by humans, so when humans compare black and white, `0.189` linear brightness lies perceptually between them.
[There](https://github.com/tobspr/GLSL-Color-Spaces/blob/master/ColorSpaces.inc.glsl) are formulas for sRGB and other colourspace transformations.

> Surprisingly, brightness and contrast did not change.

Maybe they deliberately do it in non-linear light because people who increase the contrast may want to do it in a perceptual way.
As far as I know, making an image "brighter" by a specified amount sometimes may mean that the perceived brightness should be adjusted by this amount and not the linear brightness. For example, increasing the brightness of a function plot colour should usually be done perceptually.

Some time ago I created [a mod where white cuboids are shaded](https://github.com/HybridDog/nodebox_creator/blob/fc611710f205bcad6608a25166731a08570fdd94/init.lua#L54) with the colorize texture modifier (this probably stopped working correctly in the latest Minetest due to object shading). When the shading should halve the brightness in a linear way, it is important to know that colorize is applied in sRGB colourspace.
Calculating texture modifiers in a linear colourspace has some disadvantages: The colourspace transformations may require many computations and all the texture modifiers are applied on the CPU, so it could slow down the texture generation, more code would need to be implemented just for the texture modifier colourspace transformations, and it may be somewhat backwards-incompatible (visual changes of existing nodes) unless the colourspace can be configured.
In my opinion it is important that the colour space in which the texture modifiers are applied is mentioned in the lua_api.

Treer commented:
> Calculating texture modifiers in a linear colourspace has some disadvantages: The colourspace transformations may require many computations and all the texture modifiers are applied on the CPU, so it could slow down the texture generation, more code would need to be implemented just for the texture modifier colourspace transformations, and it may be somewhat backwards-incompatible 

Yeah, all the existing texture modifiers operate in the perceptual sRGB colour space (with the possible exception of operations that get passed to Irrlicht, like scaling), and the new screen modifier is a complement of the existing multiply modifier, and the overlay and hardlight modifiers should work the same way.

> it is important that the colour space in which the texture modifiers are applied is mentioned in the lua_api.

Separate PR? I agree, but I don't know where it should go.

HybridDog commented:
> Question: If this was merged, how would a mod know that a texture modifier was available?

Texture modifiers are applied client-side. I assume with the version numbers returned by `minetest.get_player_information`, it can be tested if the client Minetest version is 5.3.0-dev or later. If the version is too old, the player would probably see texture loading error messages but can still play around.

> Separate PR? I agree, but I don't know where it should go, or how Irrlicht does it.

There is a configuration option in Irrlicht for gamma-correct texture loading, which is not enabled in Minetest.
The Irrlicht operations are probably all done directly in sRGB space. 
I have opened a PR for this now:  #10112

sfan5 commented:
Haven't reviewed the code yet but: How do `[overlay` and `[hardlight` deal with mismatching texture dimensions?
Ideally it would scale the textures to match so that usage of texture packs does not give unexpected results.

Treer commented:
I followed the way that `[mask` was is doing it, and now that you point it out, I don't think that is scaling the textures and I think they might explain some problems I've had in the past with texture modifiers.

I agree the textures should be scaled to match, should mask be fixed as well?

sfan5 commented:
`[mask` is a example of what <b>not</b> to do because it doesn't do scaling, it's essentially useless if you are working with different size textures.

> Perhaps a `[maskscaled` which invokes the same apply_mask() call?

It would be overall nicer if this could be added to `[mask` but I fear backwards compatibiliy does not make this possible (or maybe it does?). So yes, that'd be a good idea.

Treer committed:
Include texture upscaling support in Overlay and Hardlight

Treer commented:
The `[mask` question got easier - I've left it alone because if the user sets their Texturing mode to be bilinear or trilinear then the scaling interpolates values, which isn't suitable for a binary AND operation.

`[overlay` and `[hardlight` have been updated to deal with mismatching texture dimensions

hecktest commented:
You could make `^[alphamask` which simply takes a texture's luminance and multiplies the previous texture's alpha by it. Much more useful than `^[mask` and it can be scaled.

And if it wouldn't be too much trouble, consider adding blend modes that replace color or luma independently.

And because I'm on a suggestion spree already, might as well: channel extraction. (take a map's red, green or blue as grayscale)

Treer commented:
I'm not adverse to implementing more modifiers in this PR provided it keeps things simple for the reviewers - i.e. reviewing a bunch all at once when your head is in the texture-modifier space, and provided it's not someone's goal to keep the list of modifiers as minimal as possible, thus jeopardizing the PR. However I'm not a good judge of what else is needed.

> channel extraction. (take a map's red, green or blue as grayscale)

Will this already satisfy that requirment:
`texture.png^[multiply:#FF0000^[hsl:0:-128` // [multiply removes all but one channel, then [hsl desaturates it to get greyscale

long texture pipelines aren't ideal, but without knowing intended use cases it's hard to know if a separate modifier is desirable. A dedicated one would certainly be faster.

> consider adding blend modes that replace color or luma independently.

This is intriguing, could you be more specific?

I interpret this as modifiers which would take two textures, convert them into YCbCr (the colorspace jpeg uses), then
* [replaceluma would replace the the Y of one texture with the Y from the other
* [replacechroma would replace the the CbCr channels of one texture with the CbCr from the other

I can see how that could be used to recolour textures, and that it can do things that `[hsl` couldn't, but what situations would you need to swap the chroma or luma where changing hue in a blanked fashion isn't sufficient? I'm not suggesting it's unnecessary, I can think of at least one use, just trying to learn things this may unlock.

hecktest commented:
Right, your channel extraction trick almost works; problem is that the L value alone from just R, G or B will not be normalized. Channel extraction was just something I thought of that would possibly be useful to feed into the other suggested modifiers, as a "false luminance". It's not that important.

Color or luma replacement would be performed on the left hand side texture (or whole stack), using another, right hand side texture. The idea is to replace only color or only luminance of one operand with another. (the actual color space this is done in doesn't matter, as long as it perceptually looks 'right'; this can just be HSL's L and HS if it looks good) The two operations are complements of one another. The corresponding Photoshop blend modes are Color/Luminosity.

The use case for replacing color is that you can create cheap, very compressible "palette textures" that can assign different colors to different parts of the image, with local hue changes. (Take an apple's luminance, a nyan trail's color, and you have rainbow apples!)
Luminance replacement would just do the same thing with swapped operands, which can sometimes be convenient. (like when you've assembled your color map from user provided data with something like ^[combine, and now just want to give it the luminance values as a final step)
Or maybe parentheses and escaping are good enough to not need this?

I want to hear what the maintainers have to say about adding more things to this PR; whether it's better to have one PR about this or two smaller ones. I can implement the requested modes myself, or review and test them here. As long as we get them in.

One other, very trivial blend mode that we're missing is addition. Shouldn't be a problem for anyone.

KaylebJay commented:
Woah, quality PR. Well done.

Treer commented:
Is it worth adding a `[rect:<w>x<h>:<r>,<g>,<b>,<a>` modifier which creates a texture of the given size with the given color and optionally alpha?

I often encounter uses for this - colored particles, edge textures, a canvas for `[combine`, anywhere you see a mod with a 1x1 pixel .png etc, and I'm unaware of a proper way to do it. 

It seems like it wouldn't be much more code, and this PR doesn't seems to be at risk of being reviewed any time soon, so it probably wouldn't hold that up.

Treer committed:
Add [rect texture modifier

kilbith commented:
Can we have a HSL to RGB helper added to the Lua API?

appgurueu commented:
See also https://github.com/minetest/minetest/issues/11588. I'd like a redo of texture modifier syntax; we should probably refrain from adding too much modifiers before that.

appgurueu commented on doc/lua_api.txt:
```diff
@@ -617,6 +640,64 @@ Result is more like what you'd expect if you put a color on top of another
+
+`<saturation>` and `<lightness>` are optional, and both from -100 to +100
+
+#### `[contrast:<contrast>:<brightness>`
```
brightness should IMO be an optional argument of `brighten` instead of mushed in here

Treer replied:
They are not mushed together here arbitrarily. Brightness and contrast are the two knobs of the same operation, they are together here for the same reason they are together in gimp and photoshop.

Changing the slope of an adjustment line/curve changes contrast, and changing its y-intercept changes brightness. Changing them together instead of running two separate adjustment operations reduces rounding/quantization error and prevents any clipping from applying operations in a bad order. 

Usually you want the contrast to change automatically as you change brightness to avoid clipping, so conventional brightness *is* a contrast operation. But at this point I'm not sure what the "correct" UI is for the two knobs (or whether there even is a correct way), my implementation treats your contrast value as how much you wish to rotate the adjustment line, and that rotation is applied on top of any slope given to the adjustment line by the change in brightness (if one is specified).

SmallJoker commented on doc/lua_api.txt:
```diff
@@ -617,6 +640,64 @@ Result is more like what you'd expect if you put a color on top of another
+Hard light combines Multiply and Screen blend modes. Light parts of the 
+`<file>` texture will lighten (screen) the base texture, and dark parts of the
+`<file>` texture will darken (multiply) the base texture. This can be useful
+for applying embossing or chiselled effects to textures. A Hard light with the
+same texture looks like an S-curve.
```
What does S-curve mean? I don't see any difference.
![grafik](https://user-images.githubusercontent.com/1497498/132958744-19339d11-c388-4e5c-9df7-a0c97a512569.png)

Left: `testimage.png^[hardlight:testimage.png`
Right: `testimage.png`

Treer replied:
I'll have to investigate, they should look noticeably different.

[This page show an s-curve on an image](https://www.lightxapp.com/curves-tool-in-photo-editing/): the lights should get lighter and the darks should get darker

Treer replied:
It works, phew! It's just hard to see in that test image.

I've added some arrows where the lights are noticeably lighter and the darks are noticeably darker, and demo'd it with the image that webpage used.
![hardlight screenshot](https://user-images.githubusercontent.com/6390507/132973744-200e1fdc-aa3d-4463-9e46-5cb94ab9a59e.jpg)

Because this is a better way to increase contrast than with the contrast modifier, I think it could be added as a note in the contrast modifier's documentation.

Treer replied:
I've added the note to the `[contrast` documentation in [fb393c1](https://github.com/minetest/minetest/pull/10100/commits/fb393c1107f2e8d713943f3fbd3194da9bc21415#diff-da915d7065166fede89169efbfe67ec7016a1458fa6a553443039f2f1ef7a779R690-R693)

and also hopefully made it less cryptic in the hard light [desciption](https://github.com/minetest/minetest/pull/10100/commits/fb393c1107f2e8d713943f3fbd3194da9bc21415#diff-da915d7065166fede89169efbfe67ec7016a1458fa6a553443039f2f1ef7a779R712-R714)

SmallJoker reviewed (commented):
Tested the demo mod a bit, and the functions seem to work as expected.

It takes a quite long time to initialize all textures, hence I did some rudimentary measurements by loading each file individually, and subtracting the base time it took for `joinplayer` to be executed. For comparison:

`contrast`: 19 ms (49 nodes, 0.95 s)
`colorizehsl`: 62 ms (69 nodes, 4.27 s)
`hsl`: 65 ms (69 nodes, 4.48 s)

Though I doubt there can be changed much. It's complicated operations after all.

Treer commented on doc/lua_api.txt:
```diff
@@ -567,6 +567,19 @@ Example:
 Creates an inventorycube with `grass.png`, `dirt.png^grass_side.png` and
 `dirt.png^grass_side.png` textures
 
+#### `[rect:<w>x<h>:<color>`
```
Fill is a much better name.

What would the expected behaviour of a position offset be when the source image is smaller than the rect?
What would the expected image-scaling behaviour be if the position offset is negative?

Treer replied:
I was doing some work with Image Magick last night, it has its own image manipulation language and uses 'draw rectangle'. This had me thinking `[drawrect` would also be better than `[rect`, and you could say "^[draw... functions don't scale the base image". But since there'll probably never be anything like [drawline or [drawcirc I'm happy to keep it simple with [fill

Treer replied:
renamed to fill in 8252840
optional position offset added in e211ce7

Treer commented on doc/lua_api.txt:
```diff
@@ -617,6 +640,64 @@ Result is more like what you'd expect if you put a color on top of another
+
+`<contrast>` and `<brightness>` are both values from -127 to +127.
+
+`<brightness>` is optional.
```
It says that in the previous line, should it be mentioned for each one, like:
```
`<contrast>`  is a value from -127 to +127.

`<brightness>` is optional, and a value from -127 to +127.

Treer replied:
Changed in [fb393c1](https://github.com/minetest/minetest/pull/10100/commits/fb393c1107f2e8d713943f3fbd3194da9bc21415#diff-da915d7065166fede89169efbfe67ec7016a1458fa6a553443039f2f1ef7a779R686-R688)

Treer commented on doc/lua_api.txt:
```diff
@@ -617,6 +640,64 @@ Result is more like what you'd expect if you put a color on top of another
+GIMP's "Brightness-Contrast" feature but allows brightness to be wound 
+all the way up to white or down to black.
+
+`<contrast>` and `<brightness>` are both values from -127 to +127.
```
Cast a critical eye at [the 4 lines of code at line 2365](https://github.com/minetest/minetest/pull/10100/files#diff-955e227b6f551c421062f9038477a5f8d4cedf669d96bcefd22e83df5f451df0R2365-R2368)

I can explain my thinking, but that doesn't mean it's best. There are two issues.

I feel that raising brightness and lowering brightness should be symmetric operations, and that wouldn't be the case if it lowers in n/128 fractional intervals but raises in n/127 fractional intervals. However I can't think of any real-world cases where this matters because raising or lowering brightness also lowers contrast, so it's not like you could use it to undo an earlier brightness operation by doing it again with a negative value (unless doing some clever math with the contrast argument).

(The args aren't even signed bytes, the range could be -1000 to +1000, but such accuracy would be misleading with 24bit RGB values. I was probably influences by the UI choices of paint programs and other texture modifiers)

The other issue is how I'm avoiding infinite slope, perhaps a better solution to line 2367 might be
```
double norm_c = core::clamp(contrast,   -127, 127) / 127.0;
if (norm_c == 1.0)
    norm_c = 0.999;

that way a slope of zero is possible (make everything the same grey), but these extremes in a contrast operation don't seem very useful, or visually different from what the existing code does.

Treer commented on src/client/tile.cpp:
```diff
@@ -1983,6 +2183,230 @@ static void apply_multiplication(video::IImage *dst, v2u32 dst_pos, v2u32 size,
+		double dst_min = std::fmin(dst_r, std::fmin(dst_g, dst_b));
+		double delta = dst_max - dst_min;
+
+		double dst_h, dst_s, dst_l; // hue, saturation, lightness
```
looks like a linter issue: dst_max is either dst_r, dst_g, or dst_b, so there's no execution path that leaves dst_h uninitialized, but the linter can't know that.

I can initialise it to 0 to hide the warning.

Treer replied:
This code is removed by b8e41ab

Treer commented:
> It takes a quite long time to initialize all textures
>⁝
> Though I doubt there can be changed much. It's complicated operations after all.

The test images are also 256x256 (iirc), so it should be ~256 times faster on 16x16 images.

I was going to mention... irrlicht provides an HSL conversion function. I didn't use it for two reasons
* Speed - converting to HSL and back using irrlicht would add two function calls per pixel, plus (iirc) some reformating of the RGB data to match what function expects.
* I didn't want to contribute to the coupling of the client to irrlicht, which was dead at the time - this reason is not very relevant today with the irrlicht fork.

I raise this because I don't think the irrlicht algorithm gives exactly the same results as the one I used, they seem to be *roughly* the same (change hue by this much and texture goes purple using either version), but I don't think the colour values are exactly the same. Not sure if it different roundings from the different implementations or if the algorithmic approaches are more fundamentally different.

After the irrlicht fork I was thinking "C is fast, perhaps just standardise on calling irrlicht's implementation", but you measured it and it's not that fast. I'd ask if you're running on a potato but load times are noticeable on mine too.

Treer committed:
Replace RGB<->HSL conversion code with Irrlicht's implementation

Colorise might be slightly slower this way since Irrlicht will calculate hue before Colorise throws hue away.
The HSL code this commit removes gives the same visual result as Irrlicht's HSL routines, but implementation differences means color component values may sometimes differ by 1, or in very rare cases 2 (out of 255). Using Irrlicht's routines means hsl texture modifier operations will exactly match the values of any HSL functionality that might one day be exposed to lua.

Code was initially not using Irrlicht to avoid coupling the client to Irrlicht further, but this is no longer relevant today with Minetest having its Irrlicht fork. Using Irrlicht's routines keeps tile.cpp simpler

Treer committed:
rename [rect to [fill

Treer committed:
incorporate pull-request feedback in lua_api.txt

Treer committed:
Add optional position offset to ^[fill

Treer commented:
I've rebased the changes, and
- Removed my own HSL conversion code, replacing it with Irrlicht's. This might make colorizehsl slightly slower as it prevents a small optimization.
- renamed `[rect` to `[fill` and added an optional position offset to it. `^[fill` will now never scale a base image it match its size or scale itself to match a base image size, it never performs any scaling. If you want a fill operation scaled to the base image you can use `[colorize`
- Attempted to improve some lua_api.txt entries based on SmallJoker's feedback

The [texturemodifiertest](https://github.com/Treer/texturemodifiertest) repo has been updated for the new features/changes.

Zughy commented:
@Treer this should be rebased again, I'm sorry..

Treer committed:
make gcc_5 happy

Later gcc versions are happy, but not the gcc_5 build:

/home/runner/work/minetest/minetest/src/client/tile.cpp:2233:14: error: ‘fmod’ is not a member of ‘std’
    hsl.Hue = std::fmod(hsl.Hue + hue, 360);
/home/runner/work/minetest/minetest/src/client/tile.cpp:2323:21: error: ‘fabs’ is not a member of ‘std’
  double slope = 1 - std::fabs(norm_b);
/home/runner/work/minetest/minetest/src/client/tile.cpp:2327:17: error: ‘atan’ is not a member of ‘std’
  double angle = std::atan(slope);
/home/runner/work/minetest/minetest/src/client/tile.cpp:2331:10: error: ‘tan’ is not a member of ‘std’
  slope = std::tan(angle);

camera.cpp still uses std:fabs() and doesn't cause any errors, I think because it uses '#include <cmath>', so that's another possible solution.

Treer commented:
> @Treer this should be rebased again, I'm sorry..

Rebase complete

sfan5 commented on doc/lua_api.txt:
```diff
@@ -625,14 +648,24 @@ which it assumes to be a tilesheet with dimensions w,h.
+
+`<hue>` should be from -180 to +180
+
+`<saturation>` and `<lightness>` are optional, and both from -100 to +100
```
are these relative or absolute?

Treer replied:
the new text is using the degree symbol (°).

My editor thinks this file is unix utf-8 format, and I saw there's a PR/issue to switch it to a .md file, so it should be ok, but thought I'd call it out.

sfan5 commented on doc/lua_api.txt:
```diff
@@ -641,6 +674,71 @@ Result is more like what you'd expect if you put a color on top of another
+Applies an Overlay blend with the two textures, like the Overlay layer mode
+in GIMP. Overlay is the same as Hard light but with the role of the two
+textures swapped, see the `[hardlight` modifier description for more detail
+about these blend modes.
```
So `a.png^[overlay:b.png` is the same as `b.png^a.png`?
Not sure I get this

Edit: Reading on, yeah nope it isn't. Could this be explained better for people who are not familiar with anything?

Treer replied:
>> like the Overlay layer mode in GIMP
>
> So a.png^[overlay:b.png is the same as b.png^a.png?

later you suggest leaving out comments like "like XXX in GIMP", but the above confusion along with appgurueu's [suggestion that brightness should be seperated](https://github.com/minetest/minetest/pull/10100#discussion_r704519671) makes me think I'm not hammering on this drum hard enough, and it's not merely for lua modders to understand.

"like XXX in GIMP" could well be the wrong way to say it. What I'm trying to convey is there's a time-tested well thought out standard set of colour operations which are universal across CSS and every graphics program (GIMP, Photoshop, Affinity, Paint.Net etc. - all of them), they do it that way for smart reasons, and consequently they're also now the colour operations artists are familiar with. This PR is attempting to expose some of those operations in the modding API.

My implementations may leave room for improvement, e.g. someone might think [linear light](https://github.com/minetest/minetest/pull/10100#issuecomment-650471088) is worthwhile, or wonder why I used a tan call in apply_brightness_contrast(), or suspect that there's a bug in irrlicht's HSL code related to increasing saturation of greys (I suspect this, but haven't yet investigated). In each case it should be clear that unlike other modifiers, this code doesn't define what these texture modifiers are supposed to do, their definitions and answers to what/why/wtf are outside of minetest and in the wider industry. 

You're right that it's not clear, but I might not be the right person to make it clearer. I'll have to think more.

Treer replied:
perhaps I could change the first line to

`Applies [an Overlay blend](https://en.wikipedia.org/wiki/Blend_modes#Overlay) with the two textures, like the Overlay layer mode`

Hard to read as a text file but it might be a markdown file one day

sfan5 commented on src/client/tile.cpp:
```diff
@@ -537,6 +537,34 @@ static void apply_colorize(video::IImage *dst, v2u32 dst_pos, v2u32 size,
+
+// Adjust the brightness and contrast of the base image. Conceptually like
+// "Brightness-Contrast" in GIMP but allowing brightness to be wound all the
+// way up to white or down to black.
```
I think you could leave the "like XXX in GIMP" comparisons out of the function docs here

sfan5 commented on src/client/tile.cpp:
```diff
@@ -537,6 +537,34 @@ static void apply_colorize(video::IImage *dst, v2u32 dst_pos, v2u32 size,
+// Hue should be from -180 to +180, or from 0 to 360.
+// Saturation and lightness are both from -100 to +100
+static void apply_hue_saturation(video::IImage *dst, v2u32 dst_pos, v2u32 size,
+		int hue, int saturation, int lightness, bool colorize);
```
Not that there's a real reason for this but graphics code usually uses `s32` for numbers

sfan5 commented on src/client/tile.cpp:
```diff
@@ -1060,43 +1088,54 @@ static std::string unescape_string(const std::string &str, const char esc = '\\'
+			createImage(video::ECF_A8R8G8B8, dim2);
+		img1->copyToScaling(scaled_image);
+		img1->drop();
+		img1 = scaled_image;
+
+	}
+	else {
```
(code style generally mandates this)
```suggestion
		img1 = scaled_image;
	} else {
```

sfan5 commented on src/client/tile.cpp:
```diff
@@ -1265,6 +1304,44 @@ bool TextureSource::generateImagePart(std::string part_of_name,
+			video::IImage *img = driver->createImage(video::ECF_A8R8G8B8, dim);
+			img->fill(color);
+
+			if (baseimg == NULL) {
```
(either use `nullptr` or avoid it altogether like this)
```suggestion
			if (!baseimg) {
```

sfan5 commented on src/client/tile.cpp:
```diff
@@ -1824,6 +1912,109 @@ bool TextureSource::generateImagePart(std::string part_of_name,
+			sf.next(":");
+			u32 hue = mystoi(sf.next(":"), -180, 180);
+			u32 saturation = sf.at_end() ? 0 : mystoi(sf.next(":"), -100, 100);
+			u32 lightness  = sf.at_end() ? 0 : mystoi(sf.next(":"), -100, 100);
```
-100 does not fit into an u32, did you test this case?

Treer replied:
The [textureModifierTest mod](https://github.com/Treer/texturemodifiertest) tests all of these texture modifiers.

![negative-hsl](https://user-images.githubusercontent.com/6390507/179347571-58a58bd2-37ac-4e64-bc80-436526524d53.png)

I guess C++ and two's complement just don't care whether ints are signed or not :)
It's fixed now, thanks.

That test texture isn't great for testing saturation increases as it's already quite saturated, but I noticed raising sauturation turned the grey box in the centre red, which seemed odd, so I tried with the fantasy image.

It's doing what you expect for colourful parts of the image but the quality in grey parts of the image when increasing the saturation is so bad that I wonder if there's a bug in irrlicht's HSL implementation. The two images in the bottom left are made my Affinity, GIMP gives even nicer results but I think its scale just isn't going as stupidly high.

I think this will be a separate issue, but noting it here anyway

![hsl_saturation_irrlicht](https://user-images.githubusercontent.com/6390507/179348651-b1c8d196-31ad-4e46-87d9-8be5a725faa6.png)

sfan5 commented on src/client/tile.cpp:
```diff
@@ -1824,6 +1912,109 @@ bool TextureSource::generateImagePart(std::string part_of_name,
+			Strfnd sf(part_of_name);
+			sf.next(":");
+			u32 contrast = mystoi(sf.next(":"), -127, 127);
+			u32 brightness = sf.at_end() ? 0 : mystoi(sf.next(":"), -127, 127);
```
same type limit issue here

sfan5 commented on src/client/tile.cpp:
```diff
@@ -1990,6 +2181,177 @@ static void apply_multiplication(video::IImage *dst, v2u32 dst_pos, v2u32 size,
+	double norm_s = core::clamp(saturation, -100, 100) / 100.0;
+	double norm_l = core::clamp(lightness,  -100, 100) / 100.0;
+	double multiply_s = norm_s < 0 ? norm_s + 1.0 : 1.0 - norm_s;
+	double multiply_l = norm_l < 0 ? norm_l + 1.0 : 1.0 - norm_l;
```
should be floats, since `SColorf` is too

Treer committed:
PR feedback improvements

Treer committed:
Improve saturation scale in ^[hsl and ^[colorizehsl

Changes how saturation adjustments are applied, for a noticeable quality improvement when raising saturation of images containing grays or muted colors. However, this necessitates allowing saturation percentages above 100.

The solution comes from gimp, which used to have the same issue
https://github.com/GNOME/gimp/blob/6cc1e035f1822bf5198e7e99a53f7fa6e281396a/app/operations/gimpoperationhuesaturation.c#L139-L145=

Treer commented:
This saturation bug noticed [above](https://github.com/minetest/minetest/pull/10100#discussion_r922656367)... it turns out gimp used to have the very same issue, and someone awesome left [a handy dandy comment in gimp's code base](https://github.com/GNOME/gimp/blob/6cc1e035f1822bf5198e7e99a53f7fa6e281396a/app/operations/gimpoperationhuesaturation.c#L139-L145=) describing it along with the fix.

So I've copied gimp's solution. This has meant that the saturation value in ^[hsl and ^[colorizehsl needs to be able to go above +100, so I've updated the docs accordingly. It can be still be thought of as a percentage, but 100% now means "double the saturation" rather than "every pixel has 100% saturation". I've capped it at 1000%, but that's pretty arbitrary, chosen because my instinct is that normal uses will be under 400% and 1000 is nice and round.

![hsl_saturation_fixed](https://user-images.githubusercontent.com/6390507/179360164-c98a2219-9c2c-4ccb-bf89-2b7f8a210aaf.png)
Isn't it beautiful :D
(that's original at the top, with a saturation of +30, +60, and +90 on the second row,. and +200, +300, and +1000 on the bottom row)

Treer commented:
I'm not sure which party is supposed to click "Resolve conversation", but I think all the PR feedback has been implemented.

sfan5 reviewed (approved):
Should be good now.

sfan5 commented:
Note that this might need a tiny bit of rebasing if #12514 is merged first.

Treer committed:
Update to work with source image tracking (PR-12514)

https://github.com/minetest/minetest/pull/12514 "Source image tracking in tile generation" (commit adf95601) adds a new parameter to generateImage()

appgurueu commented:
Are you sure that you've compiled this properly? The head of this is https://github.com/minetest/minetest/pull/10100/commits/92d4fbb5d47079b59132becc72d0412bc7a6359b, not f4429c03a as in your screenshot.

kilbith commented:
Code have extras `+` that should be removed.

Treer commented:
I'm not sure what it's supposed to look like, or what orionflora_tree_bark.png contains, but changing the hue of a grey and black texture won't change the texture, as the greys have no color in them.

colorizehsl would give the greyscale texture an adjustable hue, but the saturation is set to 0, which removes all color and makes the texture greyscale

Treer commented:
Try raising the saturation on the colorizehsl

sfan5 commented:
I don't know if any other dev plans on reviewing this soon but you will have to rebase it, see #13438

Treer commented:
I do intend to rebase it again, and look into how grays are supposed to colourize, but life can get in the way.

Is there some urgency now?

Treer committed:
Fix colorizehsl algorithm

Gray pixels weren't colorizing correctly, so the colorize code is replaced, using gimp's colorize as a reference implementation.
https://docs.gimp.org/2.10/en/gimp-tool-colorize.html
https://github.com/GNOME/gimp/blob/6cc1e035f1822bf5198e7e99a53f7fa6e281396a/app/operations/gimpoperationcolorize.c#L234-L255

This changes the saturation parameter to have a range of 0 to 100%

Treer commented:
colorizehsl has been fixed, using [GIMPs implementation](https://github.com/GNOME/gimp/blob/6cc1e035f1822bf5198e7e99a53f7fa6e281396a/app/operations/gimpoperationcolorize.c#L234-L255) as the reference iplementation. ([GIMP docs](https://docs.gimp.org/2.10/en/gimp-tool-colorize.html))

This changes the saturation parameter to have a range of 0 to 100%

![fixed colorize](https://github.com/minetest/minetest/assets/6390507/439fd95c-296b-42b1-ba4a-43adeeea5456)

Treer committed:
(minor) formatting

Use consistent formatting between colorize and hsl calculation so that similarities are kept apparent.

rubenwardy commented:
When loading the texturemodifiertest mod, I get the following errorstream message:

```
2023-05-27 16:21:43: ERROR[Main]: Invalid color: ""
2023-05-27 16:21:43: ERROR[Main]: generateImage(): Failed to generate "[fill:16x16:-30:green"
2023-05-27 16:21:43: ERROR[Main]: generateImage(): baseimg is NULL (attempted to create texture "[fill:16x16:-30:green")
```

Treer commented:
> When loading the texturemodifiertest mod, I get the following errorstream message:

That is testing that malformed strings are handled gracefully: see https://github.com/Treer/texturemodifiertest/blob/master/fill.lua#L141-L147

if you comment those lines out the error should disappear.

I wanted to log that those messages are expected and are a "pass", but iirc the user isn't in the game when the messages are generated. (It's caught me before)

rubenwardy commented:
> I wanted to log that those messages are expected and are a "pass", but iirc the user isn't in the game when the messages are generated. (It's caught me before)

I think if you log to "error" it might show that message just before the expected error message

rubenwardy reviewed (approved):
can't see anything suspicious in the code (that isn't already there) and the test mod looks good

Treer committed:
Improve colorizehsl

SColor.getLuminance() provides a more accurate perceptual luminance than using SColorHSL.fromRGB().Luminance and mirrors how gimp's implementation works.
This also removes SColorHSL.fromRGB() from the inner loop in colorizehsl, which should improve speed.

Gimp's implementation: https://github.com/GNOME/gimp/blob/6cc1e035f1822bf5198e7e99a53f7fa6e281396a/app/operations/gimpoperationcolorize.c#L240-L242

GIMP_RGB_LUMINANCE is using coefficients 0.22248840, 0.71690369, 0.06060791 (https://github.com/GNOME/gimp/blob/6cc1e035f1822bf5198e7e99a53f7fa6e281396a/libgimpcolor/gimprgb.h#L197-L203), while irrlicht is using 0.3, 0.59, 0.11 (https://github.com/minetest/irrlicht/blob/510976f13056422ff0f90a6870772f57f03b26a9/include/SColor.h#L360-L362), I believe both are correct and the coefficients GIMP is using are for linear RGB (https://en.wikipedia.org/wiki/Relative_luminance#Relative_luminance_and_%22gamma_encoded%22_colorspaces) while irrlicht's SColor is using gamma-compressed RGB

Treer commented:
I don't think that last commit will make a visually noticeable difference, it's a slight quality improvement and a speed improvement. Gory details are in the commit message.

appgurueu commented on src/client/tile.cpp:
```diff
@@ -2042,6 +2236,196 @@ static void apply_multiplication(video::IImage *dst, v2u32 dst_pos, v2u32 size,
+	f32 norm_l = core::clamp(lightness,  -100, 100) / 100.0f;
+
+	if (colorize) {
+		hsl.Saturation = core::clamp((f32)saturation, 0.0f, 100.0f);
```
Bug. This is done after normalization.
